### PR TITLE
v1: support image name parameter in compose request 

### PR DESCRIPTION
**Pull request by jkozol**
When a user makes a compose request to image builder they should be able to specify an image name that is stored alongside the compose request similar to the uuid. If no image name is specified than none will be stored or returned when the compose info is requested.

### CODE REVIEW

File: internal/v1/api.yaml [hunk 1]
```diff
@@ -433,6 +436,8 @@ components:
         request: {}
         created_at:
           type: string
+        image_name:
```
Comment (jkozol) [inline internal/v1/api.yaml:439]: By setting the image_name as an optional return value the type becomes `*string` This seems to cause issues returning the string. @Gundersanne do you have any thoughts about what I'm doing wrong here?

Comment (ondrejbudai) [reply internal/v1/api.yaml:439]: You have to return `*string` in this case. I would help more but I need to see the error and the line of code that's causing it.

File: internal/v1/server.go [hunk 2]
```diff
@@ -469,12 +476,25 @@ func (h *Handlers) ComposeImage(ctx echo.Context) error {
...
+
+	rawCR, err := json.Marshal(dbComposeRequest)
 	if err != nil {
 		return err
 	}
 
-	err = h.server.db.InsertCompose(composeResult.Id, idHeader.Identity.AccountNumber, idHeader.Identity.Internal.OrgId, rawCR)
+	var imageName string
+	if composeRequest.ImageName == nil {
+		imageName = ""
```
Comment (Gundersanne) [inline internal/v1/server.go:492]: default value of  a string is `""` so you can simplify to
```
if composeRequest.ImageName != nil {
     imageName = *....
}
```

File: internal/db/migrations/4_alter_table_add_image_name.up.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE composes ADD COLUMN image_name VARCHAR NOT NULL DEFAULT '';
```
[the remark this answers is not in the document]
Comment (jkozol) [reply internal/db/migrations/4_alter_table_add_image_name.up.sql:1]: So, when I tried leaving the default value as NULL it creates issues when mapping from the db to a golang struct. I get error messages of can't map NULL to *string. Do you how I could change the `.Scan()` so that we can map from NULL to *string. Everything I could find online recommended using packages to handle typing between the db and the golang structs.

File: cmd/image-builder-db-test/main_test.go [hunk 4]
```diff
@@ -257,7 +266,7 @@ func testCountGetComposesSince(t *testing.T) {
 	require.Equal(t, job1, composes[0].Id)
 
 	// correct ordering (recent first)
-	composes, count, err = d.GetComposes(ANR3, fortnight * 2, 100, 0)
+	composes, count, err = d.GetComposes(ANR3, fortnight*2, 100, 0)
```
[the remark this answers is not in the document]
Comment (jkozol) [reply cmd/image-builder-db-test/main_test.go:269]: This is coming from gofmt. I can make a separate commit for the style update.

File: internal/v1/api.yaml (hunk 1 above)
Comment (jkozol) [reply internal/v1/api.yaml:439]: I guess my confusion about returning *string is that wouldn't that return a pointer to the user?

File: internal/v1/server.go (hunk 2 above)
Comment (Gundersanne) [reply internal/v1/server.go:492]: Don't derefence it, if it's non-required in the response, it won't be specified.

And I think the UI can then just check, if 'name' is in the response -> use the name, otherwise fallback to the uuid

Comment (Gundersanne) [reply internal/v1/server.go:492]: I wouldn't ever dereference. Just store the ComposeEntry with a pointer directly.

File: internal/db/migrations/4_alter_table_add_image_name.up.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gundersanne) [reply internal/db/migrations/4_alter_table_add_image_name.up.sql:1]: jackc/pgx handles this just fine if you change ComposeEntry.ImageName to `*string`

**Comment by jkozol** [comment]
Ok. I made all the changes to the `*string` type for the name. Everything seems to work as expected now :)

File: internal/v1/server.go [hunk 5]
```diff
@@ -469,12 +476,18 @@ func (h *Handlers) ComposeImage(ctx echo.Context) error {
 		return err
 	}
 
-	rawCR, err := json.Marshal(composeRequest)
+	dbComposeRequest := DBComposeRequest{
```
Comment (Gundersanne) [inline internal/v1/server.go:479]: Why the separate object?  The idea is to just throw the request in there as it comes in.

File: cmd/image-builder-db-test/main_test.go [hunk 6]
```diff
@@ -110,9 +110,18 @@ func testMigration(t *testing.T) {
 
 	migrateOneStep(t) // migrate to step 3
 
+	migrateOneStep(t) // migrate to step 4
+
+	IMGNM := "MyImageName"
```
Comment (Gundersanne) [inline cmd/image-builder-db-test/main_test.go:115]: Nitpick but use regular variable naming here, just imageName

File: internal/v1/server.go (hunk 5 above)
Comment (jkozol) [reply internal/v1/server.go:479]: I made it a separate object to not store the image name in duplicate. I am happy to pass it all in inside of the request but then I don't think we need to store it in a separate column in the database.

File: cmd/image-builder-db-test/main_test.go (hunk 6 above)
Comment (jkozol) [reply cmd/image-builder-db-test/main_test.go:115]: Will update.

File: internal/v1/server.go (hunk 5 above)
Comment (Gundersanne) [reply internal/v1/server.go:479]: The reason to store the request in the db is just so users can replay/recreate it easily later. Attributes that we want to maybe index on, and in future potentially search on, and are defined in the api spec (request is just of type {}), we should store those in a proper column imo.

Comment (jkozol) [reply internal/v1/server.go]: Makes a lot of sense. Fixed.

**Review by Gundersanne — COMMENTED** [review]
Can you also add a test case in the db/migration test where you insert a compose with 'nil' as the imageName? And then verify it's nil coming out.

File: cmd/image-builder-db-test/main_test.go [hunk 7]
```diff
@@ -125,6 +139,12 @@ func testMigration(t *testing.T) {
 	_, count, err = d.GetComposes(ANR2, fortnight, 100, 0)
 	require.NoError(t, err)
 	require.Equal(t, 1, count)
+
+	// check that the image names are as expected
+	composeEntries, count, err := d.GetComposes(ANR3, fortnight, 100, 0)
+	var nilString *string = nil
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply cmd/image-builder-db-test/main_test.go:145]: Use `require.Nil` https://pkg.go.dev/github.com/stretchr/testify/require#Nil, no need for the variable here

[the remark this answers is not in the document]
Comment (jkozol) [reply cmd/image-builder-db-test/main_test.go]: @Gundersanne fixed

File: cmd/image-builder-db-test/main_test.go [hunk 8]
```diff
@@ -110,9 +110,23 @@ func testMigration(t *testing.T) {
...
+	// inserting empty image_name should succeed
+	insert = "INSERT INTO composes(job_id, request, created_at, account_number, org_id, image_name) VALUES ($1, $2, CURRENT_TIMESTAMP, $3, $4, $5)"
+	_, err = conn.Exec(context.Background(), insert, uuid.New().String(), "{}", ANR3, ORGID1, nil)
+	require.NoError(t, err)
+
 	// Verify that after migration step 3 adding a compose request to the db requires a non empty account number.
 	d, err := db.InitDBConnectionPool(connStr(t))
-	err = d.InsertCompose(uuid.New().String(), "", ORGID1, []byte("{}"))
+	err = d.InsertCompose(uuid.New().String(), "", ORGID1, &imageName, []byte("{}"))
 	require.Error(t, err)
```
Comment (Gundersanne) [inline cmd/image-builder-db-test/main_test.go:130]: Ah I only now noticed this, these 4 lines should move up to in between migration 3 and 4 :/

Comment (jkozol) [reply cmd/image-builder-db-test/main_test.go]: Moved

File: cmd/image-builder-db-test/main_test.go [hunk 9]
```diff
@@ -112,9 +112,23 @@ func testMigration(t *testing.T) {
 
 	// Verify that after migration step 3 adding a compose request to the db requires a non empty account number.
 	d, err := db.InitDBConnectionPool(connStr(t))
-	err = d.InsertCompose(uuid.New().String(), "", ORGID1, []byte("{}"))
+	err = d.InsertCompose(uuid.New().String(), "", ORGID1, nil, []byte("{}"))
```
Comment (Gundersanne) [inline cmd/image-builder-db-test/main_test.go:115]: Because this would be using current defined Insertcomposer against a migration-3 db, it needs to be transformed into a direct query:

```
insert = "INSERT INTO composes(job_id, request, created_at, account_id, org_id) VALUES ($1, $2, CURRENT_TIMESTAMP, $3, $4)"
_, err = conn.Exec(context.Background(), insert, uuid.New().String(), "{}", "", ORGID1)
require.Error(t, err)
```

Sorry should've mentioned that. Last thing and it should be good :)

Comment (jkozol) [reply cmd/image-builder-db-test/main_test.go:115]: ahh makes sense.

**Review by Gundersanne — APPROVED** [review]
LGTM, thank you!
